### ISSUE TITLE
[requests] Unexpected error: TypeError: argument of type 'NoneType' is not iterable;

### ISSUE THREAD

**Issue by Enniomon** [yt-dlp/yt-dlp#10078]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Canada

### Provide a description that is worded well enough to be understood

Opensuse leap 15.5, After update yt-dlp and python the issue accourd before update everything was fine...

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2024.05.27 from yt-dlp/yt-dlp [12b248ce6]
[debug] Python 3.11.9 (CPython x86_64 64bit) - Linux-5.14.21-150500.55.65-default-x86_64-with-glibc2.31 (OpenSSL 1.1.1l-fips  24 Aug 2021 SUSE release 150500.17.28.2, glibc 2.31)
[debug] exe versions: ffmpeg 6.1.1 (fdk,setts), ffprobe 6.1.1
[debug] Optional libraries: brotli-1.1.0, certifi-2023.07.22, requests-2.31.0, sqlite3-3.44.0, urllib3-2.0.7
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1820 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
ERROR: [requests] Unexpected error: TypeError: argument of type 'NoneType' is not iterable; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/yt_dlp/networking/common.py", line 117, in send
    response = handler.send(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/yt_dlp/networking/_helper.py", line 208, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/yt_dlp/networking/common.py", line 337, in send
    return self._send(request)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/yt_dlp/networking/_requests.py", line 329, in _send
    requests_res = session.request(
                   ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/requests/adapters.py", line 510, in send
    conn = self._get_connection(request, verify, proxies)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/yt_dlp/networking/_requests.py", line 191, in _get_connection
    return self.get_connection(request.url, proxies)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/requests/adapters.py", line 396, in get_connection
    proxy = select_proxy(url, proxies)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/yt_dlp/networking/_helper.py", line 88, in select_proxy
    if 'no' in proxies:
       ^^^^^^^^^^^^^^^
TypeError: argument of type 'NoneType' is not iterable

Latest version: stable@2024.05.27 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2024.05.27 from yt-dlp/yt-dlp)
```

**Comment by bashonly** [comment]
passing `--compat-opt prefer-legacy-http-handler` to yt-dlp might workaround the issue for now

**Comment by bashonly** [comment]
cc @coletdjnz @Grub4K

**Comment by bashonly** [comment]
upgrading `requests` to `>=2.32.2` may also resolve the issue on your end, but this is still something that needs to be solved on yt-dlp's end

**Comment by Enniomon** [comment]
Thank you for you prompt reply, but python311-requests-2.31.0-150400.6.12.1.noarch is The highest available version that opensuse repository offers

**Comment by Enniomon** [comment]
Thank you for you prompt reply, but python311-requests-2.31.0-150400.6.12.1.noarch is The highest available version that opensuse repository offers
________________________________

**Comment by bashonly** [comment]
Your yt-dlp debug header says you have `requests 2.31.0` -- but the lines given in the traceback didn't match with https://github.com/psf/requests/blob/v2.31.0/requests/adapters.py

OpenSUSE Leap 15.5 applied a security backport patch to their [`python-requests 2.31.0`](https://build.opensuse.org/package/show/openSUSE:Leap:15.5:Update/python-requests) package:
https://build.opensuse.org/projects/openSUSE:Leap:15.5:Update/packages/python-requests/files/CVE-2024-35195.patch?expand=1

Except they botched it here:
```diff
--- requests-2.31.0.orig/requests/adapters.py
+++ requests-2.31.0/requests/adapters.py
@@ -451,7 +507,7 @@ class HTTPAdapter(BaseAdapter):
         """
 
         try:
-            conn = self.get_connection(request.url, proxies)
+            conn = self._get_connection(request, verify, proxies)
         except LocationValueError as e:
             raise InvalidURL(e, request=request)
 
```

Here's what the [actual `requests` patch](https://github.com/psf/requests/commit/eeafb6ab3a54ab01cbf864d452ea30f5fdb67ce8) did:

```diff
         try:
-            conn = self.get_connection(request.url, proxies)
+            conn = self._get_connection(request, verify, proxies=proxies, cert=cert)
```

And here's what our patched `_get_connection` method expects:

```py
    # requests 2.31.0-2.32.1
    def _get_connection(self, request, *_, proxies=None, **__):
        return self.get_connection(request.url, proxies)
```

(there was a typo in the comment; this code was only intended to be reached by `2.32.0-2.32.1`, but the typo is apparently accurate now)

So since the OpenSUSE backport patch doesn't use keyword arguments, `proxies` is not passed properly to our method and defaults to `None` -- leading to the `TypeError` in the OP.

I highly doubt that OpenSUSE is the only distro doing "stable" security backport patches, so IMO the only path forward here is to bump the minimum required version of `requests`to 2.32.2 (or even 2.32.3).

---

The TL;DR:

- `--compat-opt prefer-legacy-http-handler` might be the workaround that OP needs to use

- if the compat opt doesn't work, the OP needs to either uninstall `python-requests`, OR run `yt-dlp` in a virtual environment that doesn't use their system python packages (e.g. install yt-dlp with pipx), OR use our [linux executable](https://github.com/yt-dlp/yt-dlp/releases/latest)

- Supporting requests older than 2.32.2 is bad and we need to stop doing it asap

**Comment by Enniomon** [comment]
Thank you very much , indeed  with "--compat-opt prefer-legacy-http-handler" I don't see any error in output 🙂.  Although Its not a clean solution like before upgrade but it does the work. thank you
________________________________

**Comment by coletdjnz** [comment]
Note the error isn't fatal and the request is automatically resent using urllib instead (what prefer legacy http handler compat opt does).
